### ISSUE TITLE
Build on windows

### ISSUE THREAD

**Issue by ghost** [hughperkins/distro#53]
I am making progress. Problem is, it builds LAPLACK/BLAS with GNU. But it wants to use Visual C++ to compile, which can't link to that library (that I can figure out). So I am building LAPLACK/BLAS for visual C++, and will see if I can get it to compile with Visual C++. I also got the Intel MKL library, although it sees it it won't compile because I am missing the BLAS headers. 

Which BLAS/LPLACK is recommended, there are quite a few packages? I will try to build it with visual c++, and see if I can get torch to recognize that install because it won't link to mingw compiled libraries.

**Comment by ArtificialArtist** [comment]
I see the major problem is that it requires fortran. Either I should use Intel C++/Fortran, because I do not think Visual C++ can link to a mingw library.

**Comment by ArtificialArtist** [comment]
I do not mind using mingw at all, the problem is the cmake decides it wants visual C++ all the time and I don't know how to get it to only use mingw for everything./

**Comment by ArtificialArtist** [comment]
Ok, I am learning things. I see the problem is I can't get the luajit that comes with distro-cl, to understand mingw. So I am changing things to use this instead: https://github.com/clcarwin/luajit-rocks-mingw so it no longer uses Visual C++ so it will use the BLAS library.

**Comment by ArtificialArtist** [comment]
Ok, for the dev and other people who want to try it. You can get BLAS and LAPACK for msys by using pacman. You can also get an OpenCL version of that too. If you install msys on your system, you can remove most of the stuff in the build_windows.bat until you no longer need it.

**Comment by ArtificialArtist** [comment]
Ok, well it seems once you get luajit installed and working with luarocks, you don't really need distro-cl anymore because everything in it can then be got from luarocks. Once you get it to build on Msys64 it seems to work and you can then try to get cltorch and clnn to work.

**Comment by ArtificialArtist** [comment]
haha, now i see why distro-cl exists as getting torch7 to compile with msys is ... a challenge. But now that I understand how to build it with msys, I can use distro-cl which will compile torch. 

to build you type
cmake .. -G "MSYS Makefiles" -DCMAKE_INSTALL_PREFIX=/path/to/install

if it complains, tell it where the make utility lives with
--DCMAKE_MAKE_PROGRAM=/path/to/gnu-make

**Comment by hughperkins** [comment]
Nice! :)

**Comment by ArtificialArchitect** [comment]
Hi, I hope we can get this to work on Windows. Lua is important as many spanish and portuguese researchers can use it, so I hope I have helped in some way. I can not get torch7 to build, I can get distro-cl to build torch but I am doing it by trial and error and I am not at all good with Windows build.

**Comment by ArtificialArchitect** [comment]
What I hope is we can get the mingw team to help us make a pacman package for distro-cl and torch. You can get lua and luajit for msys with pacman, but you can't get rocks. And torch7 doesn't build on mingw because of some problems with the header files (there is conflict with some typedef like SSIZE_T). Now that would be cool, just you can pacman torch and distro-cl and not have to build because it is super hard. The hardest thing ever to build except tensorflow, which is even harder than torch.

**Comment by ArtificialArchitect** [comment]
You can get theano pacakage on msys for python, so I hope somehow to find a way to get pacman packages for other stuff. Theano is like black magic to me, I do not know how it can find the derivatives like that.

**Comment by ArtificialArchitect** [comment]
Well, I am not sure how important it is to spanish and portugues community but I just really hope to see Torch continue and for it to work on Windows. It is part of my research but I want it to be helpful to other people too, so they have access because machine learning is here and I think you have done good work to make alternatives even if Tensorflow has the AI bazel to build (and it is a real test you know, it is not a joke to me at all) but Torch is also something which needs to be carried forward. Good luck Mr Hugh I dub thee a knight I have the temporary power to knight people in the name of ... err.. something! Thank you and sorry for spam I am working on many projects a tthe moment.

**Comment by ArtificialArchitect** [comment]
Well it is called torch.  You know in the olympics, they start by running the torch. It doesn't really matter exactly to me how it is built, but if it computes correctly it is part of the game. The Olympics are the big games. I have to take a break from working on why it won't compile and rest.

**Comment by ArtificialArchitect** [comment]
I hope that we can make a pacman package for MSYS for torch7 and distro-cl. It does work, but it is extremely difficult to build on windows and would be much better for researchers if we could package it. Theano is available on MSYS2. Unfortunately I don't know anything about visual c++ and can not debug the problem with the build with that compiler.

**Comment by ArtificialArchitect** [comment]
Now I can give some summary of my research. AMD DRM will not activate for most older product, but you can use the free driver. But, unless it is ROCm supported you will probably only be able to use OpenCL 1.1. Intel has their own OpenCL for Linux. But, the drivers work on Windows well of course. Which is why Windows is important, many people use windows with amd chips. It's just that on Linux the DRM will not enable it anymore. It is not important to me, AMD makes gaming chips if you really need the GPU compute CUDA and ROCm, but OpenCL will definitely go on in more interesting ways.

**Comment by ArtificialArchitect** [comment]
Long ago when I first got started in this computer stuff. This nightmare business. I was just a young college kid, and was a newb. I was told to build my own computer, and I was rightfully mad. But then I decided i would, to get even. I went to the shop and took my time looking at those parts. I had a choice of processor: intel pentium 100, via chip with no floating points, and the peculiar AMD K6-2. I liked the name, Advanced Micro Systems. I wanted the Intel, but everyone had intel inside, and the AMD sounded cool to me. I then picked out a video card. The NVidia Riva. Why? It was the first 3d card to be able to display opengl in a window, without going fullscreen. I only had read about stuff like Cray computers, so I knew this NVidia stuff was going to win. So I stuck with AMD for many years, but now I can buy all these post-meltdown Intel chips, and I prefer Intel. AMD is a game system, and those games are too hardcore for me.
